include/dem/insertion_clear_and_add.h

```diff
@@ -0,0 +1,144 @@
+/* ---------------------------------------------------------------------
+ *
+ * Copyright (C) 2019 - 2024 by the Lethe authors
+ *
+ * This file is part of the Lethe library
+ *
+ * The Lethe library is free software; you can use it, redistribute
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE at
+ * the top level of the Lethe distribution.
+ *
+ * ---------------------------------------------------------------------
+ *
+ */
+
+#include <core/dem_properties.h>
+
+#include <dem/dem_solver_parameters.h>
+#include <dem/insertion_file.h>
+
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/particles/particle_handler.h>
+
+#ifndef lethe_insertion_clear_and_add_h
+#  define lethe_insertion_list_of_file_h
+
+template <int dim>
+class InsertionClearAndAdd : public Insertion<dim>
+{
+public:
+  InsertionClearAndAdd(
+    const DEMSolverParameters<dim>                  &dem_parameters,
+    const parallel::distributed::Triangulation<dim> &triangulation,
+    const std::vector<std::shared_ptr<Distribution>>
+      &distribution_object_container);
+
+  /**
+   * @brief The InsertionClearAndAdd class deletes every particle present in a
+   * certain region then inserts particles using data stored in files.
+   * This allows the insertion of any number of particles at a
+   * well-controlled location with any diameter value, translation and angular
+   * velocity. If well used, the insertion is guarantee to not caused an
+   * excessive overlap with the particles already present in the triangulation.
+   *
+   * @param particle_handler The particle handler of particles which are being
+   * inserted
+   * @param triangulation Triangulation to access the cells in which the
+   * particles are inserted
+   * @param dem_parameters DEM parameters declared in the .prm file
+   */
+  virtual void
+  insert(Particles::ParticleHandler<dim>                 &particle_handler,
+         const parallel::distributed::Triangulation<dim> &triangulation,
+         const DEMSolverParameters<dim> &dem_parameters) override;
+
+  /**
+   * @brief Find every cell that are completely and partially inside de the
+   * clearing box. For the cell completely inside the box, all the particles
+   * will be deleted. For those partially inside, fine verification is required.
+   *
+   * @param triangulation Triangulation to access the cells in which the
+   * particles deleted.
+   */
+  virtual void
+  find_in_clearing_box_cells(
+    const parallel::distributed::Triangulation<dim> &triangulation);
```

Comment:
Why could this not be something that is in the "insertion" base class? This way every insertion mechanism could use this approach and "remove" would be a flag to every insertion method.